sklearn/preprocessing/_discretization.py

```diff
@@ -279,9 +284,9 @@ def transform(self, X):
         np.clip(Xt, 0, self.n_bins_ - 1, out=Xt)
 
         if self.encode == 'ordinal':
-            return Xt
+            return Xt.astype(self.dtype)
 
-        return self._encoder.transform(Xt)
+        return self._encoder.transform(Xt).astype(self.dtype)
```

Comment:
We should pass the dtype argument to the `_encoder` which is a OneHotEncoder in fit, instead of converting the result.

Reply:
Yes, I will commit a second solution where I cast the dtype earlier in the function, inside the encode.